src/ParallelAlgorithms/Events/ObserveVolumeIntegrals.hpp

```diff
@@ -0,0 +1,202 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstddef>
+#include <pup.h>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "DataStructures/DataBox/DataBoxTag.hpp"
+#include "DataStructures/DataBox/Prefixes.hpp"
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/EagerMath/Determinant.hpp"
+#include "Domain/ElementMap.hpp"
+#include "Domain/Tags.hpp"
+#include "IO/Observer/Helpers.hpp"
+#include "IO/Observer/ObservationId.hpp"
+#include "IO/Observer/ObserverComponent.hpp"
+#include "IO/Observer/ReductionActions.hpp"
+#include "NumericalAlgorithms/LinearOperators/DefiniteIntegral.hpp"
+#include "Options/Options.hpp"
+#include "Parallel/CharmPupable.hpp"
+#include "Parallel/ConstGlobalCache.hpp"
+#include "Parallel/Invoke.hpp"
+#include "Parallel/Reduction.hpp"
+#include "ParallelAlgorithms/EventsAndTriggers/Event.hpp"
+#include "Utilities/ConstantExpressions.hpp"
+#include "Utilities/Functional.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
+
+/// \cond
+namespace Frame {
+struct Inertial;
+}  // namespace Frame
+/// \endcond
+
+namespace dg {
+namespace Events {
+template <size_t VolumeDim, typename ObservationValueTag, typename Tensors,
+          typename EventRegistrars>
+class ObserveVolumeIntegrals;
+
+namespace Registrars {
+template <size_t VolumeDim, typename ObservationValueTag, typename Tensors>
+// Presence of size_t template argument requires to define this struct
+// instead of using Registration::Registrar alias.
+struct ObserveVolumeIntegrals {
+  template <typename RegistrarList>
+  using f = Events::ObserveVolumeIntegrals<VolumeDim, ObservationValueTag,
+                                           Tensors, RegistrarList>;
+};
+}  // namespace Registrars
+
+template <size_t VolumeDim, typename ObservationValueTag, typename Tensors,
+          typename EventRegistrars =
+              tmpl::list<Registrars::ObserveVolumeIntegrals<
+                  VolumeDim, ObservationValueTag, Tensors>>>
+class ObserveVolumeIntegrals;
+
+/*!
+ * \ingroup DiscontinuousGalerkinGroup
+ * \brief %Observe the volume integrals of the tensors over the domain.
+ *
+ * Writes reduction quantities:
+ * - `ObservationValueTag`
+ * - `Volume` = volume of the domain
+ * - `VolInt(*)` = volume integral of the tensor
+ *
+ * \warning Currently, only one reduction observation event can be
+ * triggered at a given observation value.  Causing multiple events to run at
+ * once will produce unpredictable results.
+ */
+template <size_t VolumeDim, typename ObservationValueTag, typename... Tensors,
+          typename EventRegistrars>
+class ObserveVolumeIntegrals<VolumeDim, ObservationValueTag,
+                             tmpl::list<Tensors...>, EventRegistrars>
+    : public Event<EventRegistrars> {
+ private:
+  struct VectorPlus {
```

Comment:
Should we move this to `Utilities/Functional.hpp`?

Reply:
I also thought about this and am on the fence, but I think it wouldn't hurt to have it in `Utilities/Functional.hpp` since I expect we will need to add `std::vector<T>`s more than just here.